src/ModelContextProtocol/Authentication/AuthorizationHelpers.cs

```diff
@@ -0,0 +1,165 @@
+    public static async Task<ProtectedResourceMetadata> ExtractProtectedResourceMetadata(
+        HttpResponseMessage response,
+        Uri serverUrl,
+        CancellationToken cancellationToken = default)
+    {
+        if (response.StatusCode != System.Net.HttpStatusCode.Unauthorized)
+        {
+            throw new InvalidOperationException($"Expected a 401 Unauthorized response, but received {(int)response.StatusCode} {response.StatusCode}");
+        }
+
+        // Extract the WWW-Authenticate header
+        if (!response.Headers.WwwAuthenticate.Any())
+        {
+            throw new InvalidOperationException("The 401 response does not contain a WWW-Authenticate header");
+        }
+
+        // Look for the Bearer authentication scheme with resource_metadata parameter
+        string? resourceMetadataUrl = null;
+        foreach (var header in response.Headers.WwwAuthenticate)
+        {
+            if (header.Scheme.Equals("Bearer", StringComparison.OrdinalIgnoreCase))
+            {
+                var parameters = header.Parameter;
+                if (string.IsNullOrEmpty(parameters))
+                {
+                    continue;
+                }
+
+                // Parse the parameters to find resource_metadata
+                resourceMetadataUrl = ParseWwwAuthenticateParameters(parameters, "resource_metadata");
+                if (resourceMetadataUrl != null)
+                {
+                    break;
+                }
+            }
+        }
+
+        if (resourceMetadataUrl == null)
+        {
+            throw new InvalidOperationException("The WWW-Authenticate header does not contain a resource_metadata parameter");
+        }
+
+        Uri metadataUri = new(resourceMetadataUrl);
+        
+        var metadata = await FetchProtectedResourceMetadataAsync(metadataUri, cancellationToken) ?? throw new InvalidOperationException($"Failed to fetch resource metadata from {resourceMetadataUrl}");
+        if (!VerifyResourceMatch(metadata, serverUrl))
+        {
+            throw new InvalidOperationException(
+                $"Resource URI in metadata ({metadata.Resource}) does not match the server URI ({serverUrl})");
+        }
+
+        return metadata;
+    }
+
+    /// <summary>
+    /// Parses the WWW-Authenticate header parameters to extract a specific parameter.
+    /// </summary>
+    /// <param name="parameters">The parameter string from the WWW-Authenticate header.</param>
+    /// <param name="parameterName">The name of the parameter to extract.</param>
+    /// <returns>The value of the parameter, or null if not found.</returns>
+    private static string? ParseWwwAuthenticateParameters(string parameters, string parameterName)
+    {
+        // Handle parameters in the format: param1="value1", param2="value2"
+        var paramDict = parameters.Split(',')
+            .Select(p => p.Trim())
+            .Select(p => 
+            {
+                var parts = p.Split(['='], 2);
```

Comment:
Is the 2 correct here? That's going to set an upper bound on how many parts will be created, so if the string was erroneously `param1=foo="bar"`, this call will end up returning a string[] of length 2, with "param1" and "foo=\"bar\"". And then the check on the next line will see Length == 2 and presumably you'd want it to return but it won't.

If I'm understanding correctly, I think you want to remove the 2. And then you can also remove the brackets from around the char, which will make the call a bit more efficient, at least on .NET, where it won't need to allocate an array.